packages/angular/build/src/utils/index-file/auto-csp.ts

```diff
@@ -0,0 +1,322 @@
+ * https://developer.mozilla.org/en-US/docs/Web/HTTP/MIME_types#textjavascript
+ *
+ * @param mimeType a string that may be a MIME type
+ * @returns whether the string is a MIME type that is associated with JavaScript
+ */
+function isJavascriptMimeType(mimeType: string): boolean {
+  return JS_MIME_TYPES.has(mimeType.split(';')[0]);
+}
+
+/**
+ * Which of the type attributes on the script tag we should try passing along
+ * based on https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script/type
+ * @param scriptType the `type` attribute on the `<script>` tag under question
+ * @returns whether to add the script tag to the dynamically loaded script tag
+ */
+function shouldDynamicallyLoadScriptTagBasedOnType(scriptType: string | undefined): boolean {
+  return (
+    scriptType === undefined ||
+    scriptType === '' ||
+    scriptType === 'module' ||
+    isJavascriptMimeType(scriptType)
+  );
+}
+
+/**
+ * Calculates a CSP compatible hash of an inline script.
+ * @param scriptText Text between opening and closing script tag. Has to
+ *     include whitespaces and newlines!
+ * @returns The hash of the text formatted appropriately for CSP.
+ */
+export function hashScriptText(scriptText: string): string {
+  const hash = crypto.createHash(HASH_FUNCTION).update(scriptText, 'utf-8').digest('base64');
+  return `'${HASH_FUNCTION}-${hash}'`;
+}
+
+/**
+ * Generates the dynamic loading script and puts it in the rewriter and adds the hash of the dynamic
+ * loader script to the collection of hashes to add to the <meta> tag CSP.
+ *
+ * @param scriptContent The current streak of <script src="..."> tags
+ * @param hashes The array of hashes to include in the final CSP
+ * @param rewriter Where to emit tags to
+ */
+function emitLoaderScript(
+  scriptContent: SrcScriptTag[],
+  hashes: string[],
+  rewriter: RewritingStream,
+) {
+  const loaderScript = createLoaderScript(scriptContent);
+  hashes.push(hashScriptText(loaderScript));
+  rewriter.emitRaw(`<script>${loaderScript}</script>`);
+}
+
+/**
+ * Finds all `<script>` tags and creates a dynamic script loading block for consecutive `<script>` with `src` attributes.
+ * Hashes all scripts, both inline and generated dynamic script loading blocks.
+ * Inserts a `<meta>` tag at the end of the `<head>` of the document with the generated hash-based CSP.
+ *
+ * @param html Markup that should be processed.
+ * @returns The transformed HTML that contains the `<meta>` tag CSP and dynamic loader scripts.
+ */
+export async function autoCsp(html: string): Promise<string> {
+  const { rewriter, transformedContent } = await htmlRewritingStream(html);
+
+  let openedScriptTag: StartTag | undefined = undefined;
+  let scriptContent: SrcScriptTag[] = [];
+  let hashes: string[] = [];
+
+  rewriter.on('startTag', (tag, html) => {
+    if (tag.tagName === 'script') {
+      openedScriptTag = tag;
+      const src = getScriptAttributeValue(tag, 'src');
+
+      if (src) {
```

Comment:
Perhaps I am misunderstanding-- but if we do `if (!src) return;` here, they won't get to later on:

```js
if (scriptContent.length > 0) {
  emitLoaderScript();
}
rewriter.emitStartTag(tag);
```

which we want script tags that *do not* have a src to do? (Emit a loader script based on the consecutive `<script src="...">` seen so far and then continue writing as normal)

Sorry, the logic is really tricky here and involves falling through various if statements-- but this is the rewritten version because before in the version where I had more if's and else-if's to denote each of the subcases distinctly, there was a lot of repetition.